packages/analytics-js/src/app/RudderAnalytics.ts

```diff
@@ -133,6 +141,74 @@ class RudderAnalytics implements IRudderAnalytics<IAnalytics> {
     this.getAnalyticsInstance(writeKey).load(writeKey, dataPlaneUrl, loadOptions);
   }
 
+  /**
+   * A function to track page lifecycle events like page loaded and page unloaded
+   * @param preloadedEventsArray
+   * @param loadOptions
+   * @returns
+   */
+  // eslint-disable-next-line class-methods-use-this
+  trackPageLifecycleEvents(
+    preloadedEventsArray: PreloadedEventCall[],
+    loadOptions?: Partial<LoadOptions>,
+  ) {
+    const { trackPageLifecycle, useBeacon } = loadOptions ?? {};
+    const { events = [], enabled = false, options = {} } = trackPageLifecycle ?? {};
+    if (enabled) {
+      const visitId = generateUUID();
+      const pageLoadedTimestamp = Date.now();
+      if (events.length === 0 || events.includes(PageLifecycleEvents.PAGELOADED)) {
+        preloadedEventsArray.unshift([
+          'track',
+          PageLifecycleEvents.PAGELOADED,
+          { visitId },
+          {
+            originalTimestamp: new Date(pageLoadedTimestamp).toISOString(),
+            ...options,
+          },
+        ]);
+      }
+      if (events.length === 0 || events.includes(PageLifecycleEvents.PAGEUNLOADED)) {
+        if (useBeacon === true) {
+          // Register the page unloaded lifecycle event listeners
+          onPageLeave((isAccessible: boolean) => {
+            if (isAccessible === false) {
+              const visitDuration = Date.now() - pageLoadedTimestamp;
+              if (!state.lifecycle.loaded.value) {
```

Comment:
It is good that you've thought about this scenario. If the SDK is not in the loaded state, there is no point in queuing the event in the preloaded buffer. The SDK will not have a chance to even put it into local storage, let alone send it.

However, similar to this approach, when SDK is loaded, we need to promote this event ahead of all the stored events. Invoking the `track` API will by default add this event to the end of the queued events.
The last batch flushed may not include this event. We need to maximize the chance.

Reply:
> It is good that you've thought about this scenario. If the SDK is not in the loaded state, there is no point in queuing the event in the preloaded buffer. The SDK will not have a chance to even put it into local storage, let alone send it.

Addressed

> However, similar to this approach, when SDK is loaded, we need to promote this event ahead of all the stored events. Invoking the track API will by default add this event to the end of the queued events.
The last batch flushed may not include this event. We need to maximize the chance.

This is not needed, I have tested this scenario by reducing the size limit.

Reply:
Please resolve this

Reply:
In the above scenario, after testing, it was realized that the events left in the buffer were automatically cleared by the flush handler registered on page leave.
So, effectively all the events were flushed. There is no need to jumble the order of the events.